diff --git a/xtuner/v1/ray/environment/lagent/tokenize.py b/xtuner/v1/ray/environment/lagent/tokenize.py
index 41ae95ec..27074cf5 100644
--- a/xtuner/v1/ray/environment/lagent/tokenize.py
+++ b/xtuner/v1/ray/environment/lagent/tokenize.py
@@ -101,7 +101,7 @@ def tokenize(
                         f"Expected routed_experts_ref to be a base64 string, but got {type(routed_experts_ref)}"
                     )
                     ref_bytes = base64.b64decode(routed_experts_ref.encode("utf-8"))
-                    routed_experts = cloudpickle.loads(ref_bytes)
+                    routed_experts = ref_bytes
             else:
                 routed_experts = None
 
diff --git a/xtuner/v1/ray/rollout/worker.py b/xtuner/v1/ray/rollout/worker.py
index fff43245..ad128f17 100644
--- a/xtuner/v1/ray/rollout/worker.py
+++ b/xtuner/v1/ray/rollout/worker.py
@@ -12,6 +12,7 @@ from typing import Any, Callable, Dict, List, Optional, Union
 import httpx
 import numpy as np
 import ray
+from ray import ObjectRef, cloudpickle
 import requests  # type: ignore[import-untyped]
 from packaging.version import Version
 from ray import ObjectRef
@@ -584,8 +585,13 @@ class RolloutWorker(SingleAcceleratorWorker):
                         else:
                             cur_routed_experts = routed_experts
 
-                        history_routed_experts = await input_extra_info["routed_experts"]  # n, layer, expert
-                        ray.internal.free(input_extra_info["routed_experts"], local_only=False)
+                        history_routed_experts_key = input_extra_info["routed_experts"]
+                        if isinstance(history_routed_experts_key, ObjectRef):
+                            history_routed_experts = await input_extra_info["routed_experts"]  # n, layer, expert
+                        elif isinstance(history_routed_experts_key, str):
+                            history_routed_experts_key = cloudpickle.loads(history_routed_experts_key)
+                            history_routed_experts = await history_routed_experts_key
+                        ray.internal.free(history_routed_experts_key, local_only=False)
                         del input_extra_info
 
                         assert (history_routed_experts.shape[0] - 1) > 0 and history_routed_experts.shape[
diff --git a/xtuner/v1/rl/base/worker.py b/xtuner/v1/rl/base/worker.py
index 6ab83524..57a4feae 100644
--- a/xtuner/v1/rl/base/worker.py
+++ b/xtuner/v1/rl/base/worker.py
@@ -7,6 +7,7 @@ from pathlib import Path
 from typing import Dict, Iterable, List, Sequence, TypeAlias, TypedDict, cast
 
 import ray
+from ray import cloudpickle
 import requests
 import torch
 import torch.distributed as dist
@@ -422,7 +423,7 @@ class TrainingWorker(SingleAcceleratorWorker):
                     )
                     out_rollout_routed_expert.append(rollout_routed_experts_tensor)
                 else:
-                    rollout_routed_expert_refs = rollout_routed_expert
+                    rollout_routed_expert_refs = cloudpickle.loads(rollout_routed_expert)
                     rollout_routed_expert = ray.get(rollout_routed_expert_refs)
                     # free obj store explicitly
                     if self.sp_mesh is None or self.sp_mesh.size() == 1:
